test/aws/FederationTest.scala

```diff
@@ -81,65 +80,72 @@ class FederationTest extends AnyFreeSpec with Matchers with JodaTimeUtils {
 
       "if no time is requested" - {
         "gives default time if we're a very long way from 19:00 local time" in withSystemTime(
-          3,
-          0
-        ) {
+          hour = 3,
+          minute = 0,
+          zoneId = Some(UTC)
+        ) { time =>
           duration(
             permission,
             None,
-            Some(DateTimeZone.UTC)
+            time
           ) shouldEqual defaultLongTime
         }
 
         "gives default time if we're after 19:00 local time" in withSystemTime(
-          21,
-          0
-        ) {
+          hour = 21,
+          minute = 0,
+          zoneId = Some(UTC)
+        ) { time =>
           duration(
             permission,
             None,
-            Some(DateTimeZone.UTC)
+            time
           ) shouldEqual defaultLongTime
         }
 
         "gives until 19:00 if we're within <max time> of 19:00 local time" in withSystemTime(
-          10,
-          0
-        ) {
-          duration(permission, None, Some(DateTimeZone.UTC)) shouldEqual 9.hours
+          hour = 10,
+          minute = 0,
+          zoneId = Some(UTC)
+        ) { time =>
+          duration(permission, None, time) shouldEqual 9.hours
         }
 
         "and no timezone is supplied, provides the default time, even near 19:00" in withSystemTime(
-          15,
-          0
-        ) {
-          duration(permission, None, None) shouldEqual defaultLongTime
+          hour = 15,
+          minute = 0,
+          zoneId = None
+        ) { time =>
+          duration(permission, None, time) shouldEqual defaultLongTime
         }
 
         "and we're quite near 19:00 with a TZ, give the remaining period" in withSystemTime(
-          15,
-          0
-        ) {
-          duration(permission, None, Some(DateTimeZone.UTC)) shouldEqual 4.hours
+          hour = 15,
+          minute = 0,
+          zoneId = Some(UTC)
+        ) { time =>
+          duration(permission, None, time) shouldEqual 4.hours
         }
 
         "and we're *very* near 19:00 with a TZ, give the remaining period" ignore withSystemTime(
-          18,
-          30
-        ) {
+          hour = 18,
+          minute = 30,
+          zoneId = Some(UTC)
+        ) { time =>
           // do we need special logic near 19:00 so people don't get pointless perms?
-          duration(permission, None, Some(DateTimeZone.UTC)) shouldEqual 4.hours
+          duration(permission, None, time) shouldEqual 4.hours
         }
 
         "uses the provided timezone to calculate the correct duration" in withSystemTime(
-          15,
-          0
-        ) {
+          hour = 15,
+          minute = 0,
+          zoneId = Some(ZoneOffset.ofHours(1))
+        ) { time =>
           duration(
             permission,
             None,
-            Some(DateTimeZone.forOffsetHours(1))
-          ) shouldEqual 3.hours
+            time
+          ) shouldEqual 4.hours
```

Comment:
The test expectation appeared to be wrong previously because the comparison is between 15:00 locally and 19:00 locally, which is 4 hours.  We appeared to be comparing local time (ie. UTC + 1 hour) with UTC.

Reply:
Let's think about what we're trying to test, here? I think we're saying the system time should be irrelevant, and the test wants to check that the timezone the user provided is being used properly. I think this comes back to the added obscurity I mentioned above, mixing the current (server) time with the timezone logic. With the test as written I'd be expcting to see a timezone argument to the duration function, or at least see a timezone being applied to the current server time. If that doesn't make sense any more then I think this test (and its wording) should change?

Reply:
Yes the way I read the test was that it's trying to verify there's nothing in the code that assumes a particular timezone.  But the way the test was previously it did seem to be comparing one timezone with another.

Reply:
Federation.duration method signature changed so that it takes a Clock.  Also changed the test title.  And the test implementation now uses a property to test arbitrary timezones.